### PR TITLE
Projects and Assignment updates

### PR DESCRIPTION
**Pull request by zwolf**
*Adds a Project model that allows differentiation between education projects that require custom workflows and subject sets and those that are static. 
*Adds controller/op for above, whereby you create a project with just a slug and the ID is grabbed from Panoptes to prevent incorrect pairings.
*Updates the Assignments controller to check the project type and only clone workflows/populate new subject sets when necessary.

@marten There are still a couple commented-out ProjectController specs that I couldn't figure out why were busted. They're supposed to work exactly like the AssignmentsController ones, but rspec wasn't matching anything I passed to the operation. Clearly it all still works as the response is correct, but this is probably worth a look. ActiveInteraction and rspec's matchers are a pretty gnarly pain to get to work together and there is probably a more clever way to test this that doesn't involve literal param matching, anyway.

### CODE REVIEW

File: app/operations/assignments/create.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 module Assignments
   class Create < Operation
-    integer :base_project_id, default: Rails.application.secrets["zooniverse_project_id"]
```
Comment (marten) [inline app/operations/assignments/create.rb]: Does this mean that API clients (i.e. the current Wildcam Labs frontend) need to pass these in? Or is that already being done? /cc @shaunanoordin

Comment (shaunanoordin) [inline app/operations/assignments/create.rb]: Pass in what now? The `project_id` and `workflow_id` ? WildCam Gorongosa Lab doesn't supply any project_ids at the moment, but if that's necessary to move forward, we will implement the change, no worries. Not sure what `workflow_id` would have to go in though, but I assume nothing would be fine?

Comment (marten) [inline app/operations/assignments/create.rb]: 593 and 338 respectively. Obviously would be different for Darién.

Comment (shaunanoordin) [inline app/operations/assignments/create.rb]: Thanks @marten ! 593 and 338 for WildCam Gorongosa production, I imagine - I'll need to find out what's needed to be done for staging. (I don't believe there's any WildCam G stuff set up there ☹️  )

Comment (zwolf) [inline app/operations/assignments/create.rb]: Yes. Project ID will be required for all requests and workflow ID is required for projects with static workflows/subject sets.

**Comment by marten** [comment]
No, I think I could send in any project ID that I want, and use that to have the education-api copy a workflow. For instance, I could do `POST /assignments?project_id=2471&workflow_id=338` and it would clone workflow 338 into the Radio Meteor Zoo project.

I can also make a new project, add the `zooniverse` user as a collaborator, and then use this API to clone some non-visible workflow into my project.

Not fully sure what the solution would be. I guess we're saying that it's acceptable to let people create workflows in education projects that opt in specifically. Maybe if we made a specific edu-api user, and changed the OAuth application to be owned by that. Then projects would have to "opt-in" to this by inviting the education-api user as a collaborator on their project. They wouldn't be able to keep secret workflows after that (because "attackers" could clone them into their own projects as described) but that might be okay?

**Comment by zwolf** [comment]
If the project doesn't exist in the eduAPI database, an error should be returned by the operation (this is missing). But I missed the fact that we need a base workflow ID for projects that require cloned workflows/custom subject sets. I'm going to update the `custom_subject_set` attribute to be an nullable integer. 

* if it's nil, a workflow param is required and used to create the assignment, nothing is changed on Panoptes
* if it's filled and it is not provided, it is used as the base workflow ID to clone on that project
* If it's filled and a workflow param is provided, an error should be raised

**Comment by zwolf** [comment]
Above is now included. I guess I see what you mean about using this API to add a Panoptes project I am not affiliated with and then then clone the workflow I specify into my own project. I could check authorize a user via Panoptes roles before creating a eduAPI project.

To be honest, though, I'm not certain that the specs are doing what they think I are. Here, for instance:
https://github.com/zwolf/education-api/blob/935e4ccff3e355e93c3e31c841405b7f17cf0818/spec/controllers/assignments_controller_spec.rb#L46
I think it's just comparing blank arrays. Going to look more tomorrow (afternoon, I guess). These AI specs have been pretty annoying so if you'd like to pair on this sometime for a bit I think it'd help.

**Comment by zwolf** [comment]
@marten I think that your security concerns re: cloning workflows into non-edu projects are addressed thusly:

https://github.com/zwolf/education-api/blob/935e4ccff3e355e93c3e31c841405b7f17cf0818/app/operations/assignments/create.rb#L65

catching the error and responding here: https://github.com/zwolf/education-api/blob/i2a/app/controllers/application_controller.rb#L6

and tested here: https://github.com/zwolf/education-api/blob/i2a/spec/operations/assignments/create_spec.rb#L87

Projects that do clone workflows do not use a `workflow_id` param, but use the `base_workflow_id` attribute on the Project as a base.

**Comment by shaunanoordin** [comment]
👍  I'm ready to make changes to WildCam Gorongosa Lab (reminder to self: `export NODE_ENV=staging` on localhost) any time you're ready to roll, @zwolf . Just confirm what needs a-changin'.

I'm assuming that, for the most part, I just need to pass in a project ID (593 for Gorongosa) when...
* Creating/Deleting/Editing Assignments
* Creating/Deleting/Editing Classrooms

I also assume that...
* there's no need to supply a `workflow_id` to the create Assignments since WCG uses the 'clone base workflow' process.
* there should be no changes needed when Viewing ("select all"/"select one", i.e. the GET command) Classrooms/Assignments

OG-WG only uses the 'GET' commands IIRC, so the only thing I'd need to update is WildCam Gorongosa Lab.

Side note: while I'm confident changes can be made, I'm not sure the `staging` version of WildCam Gorongosa Lab can be tested as easily, because if I remember, there's no staging version of WildCam Gorongos, so I'd need to do something about that first. Doy.

**Comment by zwolf** [comment]
Projects must be created to essentially whitelist them, I can take care of this bit since there's no UI for it if you'd send me the project IDs along with the IDs of the workflows they clone.

Projects have a `base_workflow_id` attribute that, when you create a new Assignment, whether it will:
* clone that workflow and create a new subject set, or
* if this value is nil, save the (required) `workflow_id` param as the Assignments `workflow_id` attribute

I didn't add anything about projects to Classrooms, since they're already tied to Panoptes user groups, so no changes there. There's no relationship between Projects and Classrooms yet, though perhaps now that we're consolidating front ends this might be useful.

Clearly this is all going to need to be tested on staging before letting it loose. However staging appears to be broken and it looks like merging here deploys to production and no one seems to know anything about the packed `staging` and `legacy` (production) servers. I'm inclined to pull this onto its own instance, but without any devops support that might be tricky.

**Pull request merged**